packages/contracts/deploy/10_framework/00_ens.ts

```diff
@@ -0,0 +1,133 @@
+import {HardhatRuntimeEnvironment} from 'hardhat/types';
+import {DeployFunction} from 'hardhat-deploy/types';
+
+import {ensDomainHash, ensLabelHash, setupENS} from '../../utils/ens';
+
+import {
+  detemineDeployerNextAddress,
+  ENS_ADDRESSES,
+  getContractAddress,
+} from '../helpers';
+
+// Make sure you own the ENS set in the {{NETWORK}}_ENS_DOMAIN variable in .env
+const func: DeployFunction = async function (hre: HardhatRuntimeEnvironment) {
+  console.log(`\nDeploying framework.`);
+
+  const {deployments, getNamedAccounts, ethers, network} = hre;
+  const {deploy} = deployments;
+  const {deployer} = await getNamedAccounts();
+
+  // Get managing DAO address.
+  const managingDAOAddress = await getContractAddress('DAO', hre);
+
+  // Prepare ENS.
+  const daoDomain =
+    process.env[`${network.name.toUpperCase()}_DAO_ENS_DOMAIN`] || '';
+
+  if (!daoDomain) throw new Error('DAO domain has not been set in .env');
+
+  console.log(
+    `Using domain of "${daoDomain}", that it is owned by the deployer ${deployer}.`
+  );
+
+  const daoNode = ethers.utils.namehash(daoDomain);
+
+  const officialEnsRegistryAddress = ENS_ADDRESSES[network.name];
+  let ensRegistryAddress;
+
+  if (!officialEnsRegistryAddress) {
+    const ens = await setupENS(deployer, daoDomain);
+    ensRegistryAddress = ens.address;
+  } else {
+    ensRegistryAddress = officialEnsRegistryAddress;
+  }
+
+  const ensRegistryContract = await ethers.getContractAt(
+    'ENSRegistry',
+    ensRegistryAddress
+  );
+
+  // Approving future `ENSSubdomainRegistrar` address
+  // by using index 2, because the next deploy will be the logic address.
+  const futureAddress = await detemineDeployerNextAddress(2, deployer);
```

Comment:
The deploy function is deploy the proxy and initializes it,
and the ens registrar check if it is approved, that is why we pre-approve

but yes, we can do it later, but then we need to deploy the contract manually in multiple step, if u we think we should go this route we can, if u think there is a big problem with predicting next address

Reply:
I get the point. have you looked into `DeterministicDeployment` of hardhat-deploy ? I don't know the risks yet though.

Reply:
In your case, there's a problem because it could happen that while you're deploying these and predicting the way you do, the same deployer outside just creates a tx. then your nonce (the time you get it with `getTransactionCount()` and while you actually call it might be mismatched). Hope that's clear.

Reply:
fair point with the nonce and the deterministic address doesn't match anymore. But for now we only deploy to goerli and can restart without an issue. So I suggest we ignore this for now. 

I also guess we won't use github actions to deploy to mainnet anyway

Reply:
a task is created

Reply:
> a task is created

Please add to the task that we should consider removing the `_ens.isApprovedForAll(nodeOwner, address(this))` from the `initialize` function in `ENSSubdomainRegistrar`.